modules/sheet/editTags.js

```diff
@@ -0,0 +1,183 @@
+const nest = require('depnest')
+const { h, Value, Struct, map, computed } = require('mutant')
+const MutantArray = require('mutant/Array')
+const concat = require('lodash/concat')
+const filter = require('lodash/filter')
+const zip = require('lodash/zip')
+const forEach = require('lodash/forEach')
+const addSuggest = require('suggest-box')
+
+exports.gives = nest('sheet.editTags')
+
+exports.needs = nest({
+  'about.obs.name': 'first',
+  'keys.sync.id': 'first',
+  'sheet.display': 'first',
+  'tag': {
+    'async': {
+      'apply': 'first',
+      'create': 'first',
+      'name': 'first'
+    },
+    'html': {
+      'tag': 'first'
+    },
+    'obs': {
+      'messageTagsFrom': 'first',
+      'allTags': 'first'
+    }
+  }
+})
+
+exports.create = function(api) {
+  return nest({ 'sheet.editTags': editTags })
+
+  function editTags({ msgId }, cb) {
+    cb = cb || function() {}
+    api.sheet.display(function (close) {
+      const tagsToCreate = MutantArray([])
+      const tagsToApply = MutantArray([])
+      const tagsToRemove = MutantArray([])
+      const tagsInput = Value('')
+
+      const myId = api.keys.sync.id()
+      const messageTags = map(
+        api.tag.obs.messageTagsFrom(msgId, myId),
+        tagId => Struct({
+          tagId: Value(tagId),
+          tagName: api.about.obs.name(tagId)
+        })
+      )
+      const filteredMessages = computed(
+        [ messageTags, tagsToRemove ],
+        (tags, removedIds) => filter(tags, tag => !removedIds.includes(tag.tagId))
+      )
+
+      const messageTagsView = map(
+        filteredMessages,
+        tag => computed(tag, t => api.tag.html.tag(t, () => tagsToRemove.push(t.tagId)))
+      )
+      const tagsToApplyView = map(
+        tagsToApply,
+        tag => api.tag.html.tag(tag, () => tagsToApply.delete(tag))
+      )
+      const tagsToCreateView = map(
+        tagsToCreate,
+        tag => api.tag.html.tag({ tagName: tag, tagId: 'new' }, () => tagsToCreate.delete(tag))
+      )
+      const stagedTags = computed(
+        [messageTagsView, tagsToApplyView, tagsToCreateView],
+        (a, b, c) => h('StagedTags', concat(a, [b, c]))
+      )
+
+      const input = h('input.tags', {
+        placeholder: 'Add tags here',
+        'ev-keyup': onInput,
+        value: tagsInput()
+      })
+
+      input.addEventListener('suggestselect', onSuggestSelect)
+
+      return {
+        content: [
+          stagedTags,
+          h('EditTags', input)
+        ],
+        footer: [
+          h('button.save', {
+              'ev-click': () => {
```

Comment:
this could just be `'ev-click': publish` (unless the order of onSave / close matters)